### PR TITLE
Explicit `undefined` return type checked similar to explicit `void` return type

### LINKED ISSUE #36288: Allow non returning functions to have contextually required `undefined` return type
## Search Terms

#36239

## Suggestion

Allow non returning functions to have contextually required `undefined` return type.

## Use Cases

- Callbacks

## Examples

```ts
declare function f(a: () => undefined): void;
f(() => { }); // error -> ok
f((): undefined => { }); // error -> ok
const g1: () => undefined = () => { }; // error -> ok
const g2 = (): undefined => { }; // error -> ok if possible
function h1() {
}
f(h1); // error -> error
function h2(): undefined { // error -> ok if possible
}
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### LINKED ISSUE #53473: TS2366: Function lacks ending return statement in function with `unknown` return type with implicit return on some code paths, despite `noImplicitReturns` turned off
# Bug Report

### 🔎 Search Terms

lacks ending

### 🕗 Version & Regression Information

- This changed between versions 5.0.2 and 5.1.0-dev.20230323

I believe [this](https://github.com/microsoft/TypeScript/pull/53092/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8L35565-R35570) caused it.

### ⏯ Playground Link

https://www.typescriptlang.org/play?noImplicitReturns=false&ts=5.1.0-dev.20230323#code/GYVwdgxgLglg9mABFApgZygCgJQC5HgDWYcA7kgN4CwAUAJAzCKYCyAhlABYB0ATm2AAmcALY5EAPkQAGbgFZsiXiigheSKLxAoA3LQC+QA

### 💻 Code

```ts
function test(): unknown {
	if (Math.random() > 0.5) return true;
}
```

### 🙁 Actual behavior

```
> tsc --noEmit

index.ts:1:18 - error TS2366: Function lacks ending return statement and return type does not include 'undefined'.

1 function test(): unknown {
                   ~~~~~~~
```

### 🙂 Expected behavior

No error, if the function return type is `unknown` I'd expect that to allow implicit return of `undefined`.

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR (hopefully) completes the work started in #53092 and subsequently modified in #53490. With this and the other PRs, checking of functions changes as follows compared to 5.0:

* In functions with an explicit or contextual return type `undefined`, expression-less `return` statements and the implicit return and the end of the function are considered to have type `undefined` (as opposed to `void`). Note that this does not apply to functions with union return types that include `undefined`.
* Functions with an explicitly specified return type `undefined` aren't required to have return statements (similar to functions with an explicitly specified return type of `void` or `any`). Note that this does not apply to functions with union return types that include `undefined`.
* The `--noImplicitReturns` option doesn't apply to functions with return types that include `undefined` (similar to functions with return types that include `void`, `any`, or `unknown`).

Fixes #36288.
Fixes #53473.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by gabritto** [comment]
@typescript-bot pack this

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster

**Comment by gabritto** [comment]
I'm confused by this example here: https://www.staging-typescript.org/play?ts=5.1.0-pr-53607-12#code/GYVwdgxgLglg9mABMOcAUBKAXIgzlAJxjAHNEAfRcAEwFNhjbrEBvAWAChFvEZhE0AQkGEQtDIgK0oIAkgBE8gNycAvp06hIsBIhKo0EAEY40uHPiKkJAXgB8eQsTKUa9RtWyIAbnBjMWdQ4IBHxEAA9EGz0DM1sHFl5+IRECMQkpGTlERURVDBUOThCwMIBPU3NHKxJ46ucKKjA6BjAmKIE69i4ePgFhUXFJaVkFZTUgA

It seems that on 5.0 with `--noImplicitReturns` we would error for explicitly annotated functions *and* for contextually typed functions containing `undefined`, but on this PR we now only error on the explicitly annotated function, is that correct? It seems a bit confusing to me.

**Comment by ahejlsberg** [comment]
Much better on the second perf suite run. Morale: Just run the tests until you get the desired result!

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
I reviewed, but let's just answer outstanding comments from @gabritto first.

**Comment by ahejlsberg** [comment]
@typescript-bot user test this inline

**Comment by jakebailey** [comment]
I'm looking into what's going on with those tests. Not sure what's going wrong; it looks green but then decides to say that something failed.

**Comment by jakebailey** [comment]
It looks like one of the user tests no longer npm installs correctly; not totally sure why we fail the whole thing due to that, though. Otherwise, reading the logs, this PR shows no change in any of those user tests.

**Comment by ahejlsberg** [comment]
@gabritto Yeah, it would be more consistent for all of them to return the error. I will change the logic to have `noImplicitReturns` only exclude functions that have exactly `undefined` as their return type.

File: src/services/codefixes/returnValueCorrect.ts [hunk 1]
```diff
@@ -51,7 +51,6 @@ const fixIdAddReturnStatement = "fixAddReturnStatement";
 const fixRemoveBracesFromArrowFunctionBody = "fixRemoveBracesFromArrowFunctionBody";
 const fixIdWrapTheBlockWithParen = "fixWrapTheBlockWithParen";
 const errorCodes = [
-    Diagnostics.A_function_whose_declared_type_is_neither_void_nor_any_must_return_a_value.code,
```
[the remark this answers is not in the document]
Comment (jakebailey) [reply src/services/codefixes/returnValueCorrect.ts:54]: For some reason we had two wordings of the same problem; check out the diagnosticMessages file.

File: tests/baselines/reference/functionsMissingReturnStatementsAndExpressions.errors.txt [hunk 2]
```diff
@@ -119,10 +123,34 @@ tests/cases/compiler/functionsMissingReturnStatementsAndExpressions.ts(135,5): e
...
-!!! error TS2355: A function whose declared type is neither 'void' nor 'any' must return a value.
+!!! error TS2355: A function whose declared type is neither 'undefined', 'void', nor 'any' must return a value.
         // Error; because `undefined | number` becomes `number` without strictNullChecks.
     }
     
+    const f30: () => undefined = () => {
+        // Ok, contextual type for implicit return is `undefined`
+    }
+    
+    const f31: () => undefined = () => {
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply tests/baselines/reference/functionsMissingReturnStatementsAndExpressions.errors.txt:134]: Those are all just regular `void` returning functions for which we already have plenty of coverage in the test suites.

File: tests/cases/compiler/functionsMissingReturnStatementsAndExpressionsStrictNullChecks.ts [hunk 3]
```diff
@@ -2,18 +2,60 @@
...
+    // Error, return type doesn't include undefined
 }
 
-async function f4(): Promise<number> {
-    // Error; return type does not include undefined
+const f20: () => undefined = () => {
+    // Ok, contextual type for implicit return is undefined
 }
+
+const f21: () => undefined | number = () => {
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply tests/cases/compiler/functionsMissingReturnStatementsAndExpressionsStrictNullChecks.ts:21]: Sure, I can fix the comments.

**Comment by ahejlsberg** [comment]
In addition to addressing @gabritto's feedback, latest commits align the rules for `--noImplicitReturns` to remove inconsistencies between functions with explicit `unknown` return types and functions with inferred `unknown` return types. Previously, only functions with explicit `unknown` return types would issue errors for implicit returns in `--noImplicitReturns` mode. Now both do.

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
A few new errors in `vscode` and `vuetify` related to the change I mention [here](https://github.com/microsoft/TypeScript/pull/53607#issuecomment-1493017136). I think these new errors are acceptable--the projects have `noImplicitReturns` enabled, the functions in question have inferred return types `unknown` (or `Promise<unknown>`), and the errors would have been reported if the functions had explicit return type annotations. It looks to me like all of the examples actually intended for the functions to return `void` but inadvertently returned values in some conditional code paths.

**Comment by ahejlsberg** [comment]
Otherwise thinks look good, so I'm going to merge this.

**Comment by stevenwdv** [comment]
Sorry for replying to an old thread, but can I ask a question about this?

Why is this not allowed for unions with `undefined`? E.g. for event listeners:

```ts
function addListener(listener: (ev: Event) => undefined | boolean) {/*...*/}
addListener(ev => {/*... (no return)*/});
// Argument of type '(ev: Event) => void' is not assignable to parameter of type '(ev: Event) => boolean | undefined'.
//   Type 'void' is not assignable to type 'boolean | undefined'.
```

I think this could be a nice use case.

**Pull request merged**